x/farm/keeper/farm.go

```diff
@@ -0,0 +1,244 @@
+	if !found {
+		return nil
+	}
+	cacheCtx, _ := ctx.CacheContext()
+	endPeriod := k.incrementFarmPeriod(cacheCtx, denom)
+	return k.calculateRewards(cacheCtx, position, endPeriod)
+}
+
+func (k Keeper) calculateRewards(ctx sdk.Context, position types.Position, endPeriod uint64) sdk.DecCoins {
+	if position.StartingBlockHeight == ctx.BlockHeight() {
+		return nil
+	}
+	startPeriod := position.PreviousPeriod
+	return k.rewardsBetweenPeriods(
+		ctx, position.Denom, startPeriod, endPeriod, position.FarmingAmount)
+}
+
+func (k Keeper) initializeFarm(ctx sdk.Context, denom string) types.Farm {
+	farm := types.Farm{
+		TotalFarmingAmount: sdk.ZeroInt(),
+		CurrentRewards:     sdk.DecCoins{},
+		OutstandingRewards: sdk.DecCoins{},
+		Period:             1,
+	}
+	k.SetFarm(ctx, denom, farm)
+	k.SetHistoricalRewards(ctx, denom, 0, types.HistoricalRewards{
+		CumulativeUnitRewards: sdk.DecCoins{},
+		ReferenceCount:        1,
+	})
+	return farm
+}
+
+func (k Keeper) updatePosition(ctx sdk.Context, position types.Position) {
+	farm, found := k.GetFarm(ctx, position.Denom)
+	if !found { // Sanity check
+		panic("farm not found")
+	}
+	prevPeriod := farm.Period - 1
+	k.incrementReferenceCount(ctx, position.Denom, prevPeriod)
+	position.PreviousPeriod = prevPeriod
+	position.StartingBlockHeight = ctx.BlockHeight()
+	k.SetPosition(ctx, position)
+}
+
+func (k Keeper) incrementFarmPeriod(ctx sdk.Context, denom string) (prevPeriod uint64) {
+	farm, found := k.GetFarm(ctx, denom)
+	if !found { // Sanity check
+		panic("farm not found")
+	}
+	unitRewards := sdk.DecCoins{}
+	if farm.TotalFarmingAmount.IsZero() {
+		// TODO: do something special?
+	} else {
+		unitRewards = farm.CurrentRewards.QuoDecTruncate(sdk.NewDecFromInt(farm.TotalFarmingAmount))
+	}
+	hist, found := k.GetHistoricalRewards(ctx, denom, farm.Period-1)
+	if !found { // Sanity check
+		panic("historical rewards not found")
+	}
+	k.decrementReferenceCount(ctx, denom, farm.Period-1)
+	k.SetHistoricalRewards(ctx, denom, farm.Period, types.HistoricalRewards{
+		CumulativeUnitRewards: hist.CumulativeUnitRewards.Add(unitRewards...),
+		ReferenceCount:        1,
+	})
+	farm.CurrentRewards = sdk.DecCoins{}
+	prevPeriod = farm.Period
+	farm.Period++
+	k.SetFarm(ctx, denom, farm)
+	return prevPeriod
+}
+
+func (k Keeper) incrementReferenceCount(ctx sdk.Context, denom string, period uint64) {
+	hist, found := k.GetHistoricalRewards(ctx, denom, period)
+	if !found { // Sanity check
+		panic("historical rewards not found")
+	}
+	if hist.ReferenceCount >= 2 { // Sanity check
+		panic("reference count must never exceed 2")
+	}
+	hist.ReferenceCount++
+	k.SetHistoricalRewards(ctx, denom, period, hist)
+}
+
+func (k Keeper) decrementReferenceCount(ctx sdk.Context, denom string, period uint64) {
+	hist, found := k.GetHistoricalRewards(ctx, denom, period)
+	if !found { // Sanity check
+		panic("historical rewards not found")
+	}
+	if hist.ReferenceCount == 0 {
+		panic("reference count must not be negative")
+	}
```

Comment:
Thank you for catching this out. Actually, the panic message seems wrong because `HistoricalRewards.ReferenceCount` is a `uint32` field and can never be negative. I'll correct the message.

Reply:
Ah, I misunderstood. That panic message prevents decrementing the reference count from `0` to negative, so the message is correct already! In other words, it doesn't check the *current* state but the *next* state.

Reply:
As the param `ReferenceCount` is already unsigned integer, it can't be negative. You're right. I see it now.